### PR TITLE
Remove uniform array move MIR passes

### LINKED ISSUE #66502: Slice/Array patterns are not correctly move checked
The following should error, but doesn't

```rust
#![feature(slice_patterns)]

fn main() {
    let a = [String::from("HELLO"), String::from("WORLD")];
    let [_, mut x] = a;
    x = String::from("RUST");
    let [ref y @ ..] = a;
    println!("{:?}", y); 
}
```

### PR DESCRIPTION
**Pull request by matthewjasper**
This PR fixes a number of bugs caused by limitations of this pass

* Projections from constant indexes weren't being canonicalized
* Constant indexes from the start weren't being canonicalized (they could have different min_lengths)
* It didn't apply to non-moves

This PR makes the following changes to support removing this pass:

* ConstantIndex of arrays are now generated in a canonical form (from the start, min_length is the actual length).
* Subslices are now split when generating move paths and when checking subslices have been moved.

Additionally

* The parent move path of a projection from an array element is now calculated correctly

closes #66502

### CODE REVIEW

File: src/test/ui/borrowck/borrowck-slice-pattern-element-loan-array.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
+//compile-flags: -Z borrowck=mir
```
Comment (Centril) [inline src/test/ui/borrowck/borrowck-slice-pattern-element-loan-array.rs]: How come?

File: src/test/ui/borrowck/borrowck-slice-pattern-element-loan-array.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+#![feature(slice_patterns)]
+
+fn nop(_s: &[& i32]) {}
+fn nop_subslice(_s: &[i32]) {}
+
+fn const_index_ok(s: &mut [i32; 4]) {
```
Comment (Centril) [inline src/test/ui/borrowck/borrowck-slice-pattern-element-loan-array.rs]: Would be good if we could split the test into one file for pass and one for fail.

**Comment by Centril** [comment]
cc also @pnkfelix @nikomatsakis @lqd

**Comment by varkor** [comment]
I won't have time to review such a large change for a little while, so reassigning.

r? @pnkfelix

**Comment by JohnCSimon** [comment]
Ping form triage 
@matthewjasper - can you please address the merge conflicts?
Thank you.

File: src/librustc_mir/borrow_check/mod.rs [hunk 3]
```diff
@@ -1614,6 +1613,46 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
+        to: u32,
+    ) {
+        if let Some(mpi) = self.move_path_for_place(place_span.0) {
+            let mut child = self.move_data.move_paths[mpi].first_child;
+            while let Some(child_mpi) = child {
+                let child_move_place = &self.move_data.move_paths[child_mpi];
+                let child_place = &child_move_place.place;
+                let last_proj = child_place.projection.last().unwrap();
+                if let ProjectionElem::ConstantIndex { offset, from_end, .. } = last_proj {
+                    debug_assert!(!from_end, "Array constant indexing shouldn't be `from_end`.");
```
Comment (pnkfelix) [inline src/librustc_mir/borrow_check/mod.rs]: (i'm inclined to make this a full `assert!`.)

File: src/test/ui/borrowck/borrowck-move-out-from-array-no-overlap.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    let [_, _y @ ..] = a;
+}
+
+fn move_out_by_const_index_end_and_subslice() {
+    let a = array();
+    let [.., _x] = a;
+    let [_y @ .., _] = a;
+}
+
+fn move_out_by_const_index_field_and_subslice() {
```
Comment (pnkfelix) [inline src/test/ui/borrowck/borrowck-move-out-from-array-no-overlap.rs]: beautiful set of tests

**Comment by pnkfelix** [comment]
Okay this looks great. If you can get around to changing the `debug_assert!` to an `assert!` before bors merges it, then that would be nice. But I'm going to go ahead and start the race anyway by r+'ing it.

**Pull request merged**